File: android/app/src/main/java/cn/kkcode/remote/ConversationPresentation.kt
package cn.kkcode.remote

/** View-only grouping. Canonical messages and live rows are never discarded. */
internal fun collapseCompletedRuns(items: List<ChatItem>, busy: Boolean): List<ChatItem> {
    val result = mutableListOf<ChatItem>()
    var pending = mutableListOf<ChatItem>()
    var started = 0L
    fun flush(last: Boolean) {
        val answer = pending.indexOfLast { it.kind == "assistant" && it.text.isNotBlank() }
        val activity = pending.filterIndexed { index, item -> index != answer && item.kind in listOf("assistant", "tool", "thinking", "review") }
        if(answer >= 0 && pending[answer].done && activity.isNotEmpty() && !(busy && last) && pending.none { it.kind == "error" } && pending.drop(answer + 1).none { it.kind in listOf("tool", "thinking", "review") }) {
            val end = pending.maxOfOrNull { it.startedAt + (it.durationMs ?: 0) } ?: 0L
            val begin = started.takeIf { it > 0 } ?: pending.map { it.startedAt }.filter { it > 0 }.minOrNull() ?: 0L
            result += ChatItem("run-${pending.first().turnId.ifBlank { pending.first().id }}", "run-summary", "", startedAt = begin,
                durationMs = if(begin > 0 && end >= begin) end - begin else null, children = activity)
            result += pending.filterIndexed { index, item -> index == answer || item !in activity }
        } else result += pending
        pending = mutableListOf()
    }
    for(item in items) {
        if(item.kind in listOf("user", "compacted")) { flush(false); result += item; started = item.startedAt }
        else pending += item
    }
    flush(true)
    return result
}

File: apps/web/src/conversation-presentation.mjs
/** Collapse only completed runs with a final answer; keep active/error traces. */
export function collapseCompletedRuns(rows, busy = false) {
  const result = [];
  let pending = [], started = 0;
  function flush(last) {
    const answer = pending.findLastIndex(row => row.type === 'assistant' && row.text);
    const activity = pending.filter((row, index) => index !== answer && ['assistant', 'tool', 'thinking', 'review'].includes(row.type));
    if (answer >= 0 && pending[answer].done !== false && activity.length && !(busy && last) && !pending.some(row => row.type === 'error') && !pending.slice(answer + 1).some(row => ['tool', 'thinking', 'review'].includes(row.type))) {
      const end = Math.max(...pending.map(row => row.finishedAt || row.updatedAt || row.timestamp || 0));
      const times = pending.map(row => row.timestamp).filter(value => value > 0);
      const begin = started || (times.length ? Math.min(...times) : 0);
      result.push({ id: `run-${pending[0].turnId || pending[0].id}`, type: 'run-summary', rows: activity, durationMs: begin && end >= begin ? end - begin : null, tools: activity.filter(row => row.type === 'tool').length });
      result.push(...pending.filter((row, index) => index === answer || !activity.includes(row)));
    } else result.push(...pending);
    pending = [];
  }
  for (const row of rows) {
    if (['user', 'compacted'].includes(row.type)) { flush(false); result.push(row); started = row.timestamp || 0; }
    else pending.push(row);
  }
  flush(true);
  return result;
}
